File: learning-core/executable/cache/python/cache.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Repository:
    data: dict[str, str]
    reads: int = 0

    def get(self, key: str) -> str:
        self.reads += 1
        return self.data[key]


class CacheAside:
    def __init__(self, repository: Repository) -> None:
        self.repository = repository
        self.cache: dict[str, str] = {}

    def get(self, key: str) -> str:
        if key in self.cache:
            return self.cache[key]
        value = self.repository.get(key)
        self.cache[key] = value
        return value

    def invalidate(self, key: str) -> None:
        self.cache.pop(key, None)

File: learning-core/executable/cache/go/cache.go
package cache

type Repository struct {
	Data  map[string]string
	Reads int
}

func (r *Repository) Get(key string) string {
	r.Reads++
	return r.Data[key]
}

type CacheAside struct {
	Repo  *Repository
	Cache map[string]string
}

func NewCacheAside(repo *Repository) *CacheAside {
	return &CacheAside{Repo: repo, Cache: map[string]string{}}
}

func (c *CacheAside) Get(key string) string {
	if value, ok := c.Cache[key]; ok {
		return value
	}
	value := c.Repo.Get(key)
	c.Cache[key] = value
	return value
}

func (c *CacheAside) Invalidate(key string) {
	delete(c.Cache, key)
}
